analyzer_test.go

```diff
@@ -118,532 +114,56 @@ func testAnalyzerBuilder(platformAPI string) func(t *testing.T, when spec.G, it
-						{"metadata.buildpack/launch.sha", "launch-sha"},
-						{"no.cache.buildpack/some-layer.sha", "some-layer-sha"},
-					} {
-						got := h.MustReadFile(t, filepath.Join(layersDir, data.name))
-						h.AssertStringContains(t, string(got), data.want)
-					}
-				})
-
-				it("does not restore layer sha files", func() {
-					h.SkipIf(t, api.MustParse(platformAPI).Compare(api.MustParse("0.7")) < 0, "Platform API < 0.7 restores layer metadata")
-
-					_, err := analyzer.Analyze()
-					h.AssertNil(t, err)
-
-					for _, data := range []struct{ name, want string }{
-						{"metadata.buildpack/launch.sha", ""},
-						{"metadata.buildpack/launch-build-cache.sha", ""},
-						{"metadata.buildpack/launch-cache.sha", ""},
-						{"no.cache.buildpack/some-layer.sha", ""},
-					} {
-						h.AssertPathDoesNotExist(t, data.name)
-					}
-				})
-
-				it("does not restore launch=false layer metadata", func() {
-					h.SkipIf(t, api.MustParse(platformAPI).Compare(api.MustParse("0.7")) >= 0, "Platform API >= 0.7 does not restore layer metadata")
-
-					_, err := analyzer.Analyze()
-					h.AssertNil(t, err)
-
-					h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "metadata.buildpack", "launch-false.toml"))
-					h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "metadata.buildpack", "launch-false.sha"))
-				})
-
-				it("does not restore build=true, cache=false layer metadata", func() {
-					h.SkipIf(t, api.MustParse(platformAPI).Compare(api.MustParse("0.7")) >= 0, "Platform API >= 0.7 does not restore layer metadata")
-
-					_, err := analyzer.Analyze()
-					h.AssertNil(t, err)
-
-					h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "metadata.buildpack", "launch-build.sha"))
-				})
-
-				when("subset of buildpacks are detected", func() {
-					it.Before(func() {
-						analyzer.Buildpacks = []buildpack.GroupBuildpack{{ID: "no.cache.buildpack", API: api.Buildpack.Latest().String()}}
-					})
-
-					it("restores layers for detected buildpacks", func() {
-						h.SkipIf(t, api.MustParse(platformAPI).Compare(api.MustParse("0.7")) >= 0, "Platform API >= 0.7 does not restore layer metadata")
-
-						_, err := analyzer.Analyze()
-						h.AssertNil(t, err)
-
-						path := filepath.Join(layersDir, "no.cache.buildpack", "some-layer.toml")
-						got := h.MustReadFile(t, path)
-						want := "[metadata]\n  some-layer-key = \"some-layer-value\""
-
-						h.AssertStringContains(t, string(got), want)
-					})
-
-					it("does not restore layers for undetected buildpacks", func() {
-						h.SkipIf(t, api.MustParse(platformAPI).Compare(api.MustParse("0.7")) >= 0, "Platform API >= 0.7 does not restore layer metadata")
-
-						_, err := analyzer.Analyze()
-						h.AssertNil(t, err)
-
-						h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "metadata.buildpack"))
-					})
-				})
-
 				it("returns the analyzed metadata", func() {
+					metadataRestorerExpectations(metadataRestorer, analyzer, expectedAppMetadata, expectedCacheMetadata)
```

Comment:
I really didn't like having to do this in multiple places, but I couldn't find a clean way to either clear expectations when I wanted to set a new one or to defer setting up the expectation until right before each test with the right context. Open to ideas though!